tests/44account_data.pl

```diff
@@ -0,0 +1,269 @@
+      })->then( sub {
+         my ( $body ) = @_;
+
+         check_one_account_data_event(
+            $body->{account_data}, "my.test.type", "rats"
+         );
+
+         Future->done(1);
+      });
+   };
+
+
+test "Account data appears in v1 /events stream",
+   requires => [ local_user_fixture() ],
+
+   check => sub {
+      my ( $user ) = @_;
+
+      Future->needs_all(
+         await_event_for( $user, filter => sub {
+            my ( $event ) = @_;
+
+            return unless $event->{type} eq "my.test.type"
+               and $event->{content}{cats_or_rats} eq "cats";
+         }),
+         matrix_add_account_data( $user, "my.test.type", {
+            cats_or_rats => "cats",
+         }),
+      );
+   };
+
+
+test "Room account data appears in v1 /events stream",
+   requires => [ local_user_and_room_fixtures() ],
+
+   check => sub {
+      my ( $user, $room_id ) = @_;
+
+      Future->needs_all(
+         await_event_for( $user, filter => sub {
+            my ( $event ) = @_;
+
+            return unless $event->{type} eq "my.test.type"
+               and $event->{content}{cats_or_rats} eq "rats"
+               and $event->{room_id} eq $room_id;
+         }),
+         matrix_add_room_account_data( $user, $room_id, "my.test.type", {
+            cats_or_rats => "rats",
+         }),
+      );
+   };
+
+
+test "Latest account data appears in v2 /sync",
+   requires => [ local_user_and_room_fixtures() ],
+
+   check => sub {
+      my ( $user, $room_id ) = @_;
+
+      setup_account_data( $user, $room_id )->then( sub {
+         matrix_sync( $user );
+      })->then( sub {
+         my ( $body ) = @_;
+
+         check_one_account_data_event(
+            $body->{account_data}{events}, "my.test.type", "cats"
+         );
+
+         check_one_account_data_event(
+            $body->{rooms}{joined}{$room_id}{account_data}{events},
+            "my.test.type", "rats"
+         );
+
+         Future->done(1);
+      });
+   };
+
+sub setup_incremental_account_data
+{
+   my ( $user, $room_id, $type, $top_animal, $room_animal ) = @_;
+
+   Future->needs_all(
+      matrix_add_account_data( $user, $type, {
+         cats_or_rats => $top_animal,
+      }),
+      matrix_add_room_account_data( $user, $room_id, $type, {
+         cats_or_rats => $room_animal,
+      })
+   );
+}
+
+test "New account data appears in incremental v2 /sync",
+   requires => [ local_user_and_room_fixtures() ],
+
+   check => sub {
+      my ( $user, $room_id ) = @_;
+
+      my ( $next_batch );
+
+      Future->needs_all(
+         setup_incremental_account_data(
+            $user, $room_id, "my.unchanging.type", "lions", "tigers"
+         ),
+         setup_incremental_account_data(
+            $user, $room_id, "my.changing.type", "dogs", "frogs"
+         ),
+      )->then( sub {
+         matrix_sync( $user );
+      })->then( sub {
+         my ( $body ) = @_;
+
+         $next_batch = $body->{next_batch};
```

Comment:
The generally-prevailing style around most tests is some shared captured variables at the top of the test block, that get set in some stage and read later. For consistency it's probably best to stick to that style.

Neither of them is overly nice - ideally we'd want to use the `async`/`await` syntax for a lot of these tests instead, but I haven't got around to implementing that yet... ;)